### PR TITLE
Post: mlr3fselect basics

### PR DESCRIPTION
**Pull request by be-marc**

### CODE REVIEW

**Comment by pat-s** [comment]
@be-marc Any rough ETA for this one?

File: DESCRIPTION [hunk 1]
```diff
@@ -65,7 +65,13 @@ Imports:
     xgboost
 Remotes:
     mlr-org/mlr3book,
+<<<<<<< HEAD
+    mlr-org/mlr3fselect,
+    mlr-org/mlr3pipelines,
+    mlr3learners/mlr3learners.randomforest
+=======
     mlr-org/mlr3extralearners
+>>>>>>> master
```
Comment (pat-s) [inline DESCRIPTION]: @be-marc FYI

Comment (be-marc) [reply DESCRIPTION]: I should not push and pack in. Thanks :)

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+if (require("data.table")) data.table::setDTthreads(1)
+set.seed(7832)
+lgr::get_logger("mlr3")$set_threshold("warn")
+```
+
+## Introduction
+
+In this tutorial, we introduce the **mlr3fselect** package by comparing feature selection methods on the titanic disaster data set.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
In this tutorial, we introduce the **mlr3fselect** package by comparing feature selection methods on the Titanic disaster data set. 
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+```
+
+## Introduction
+
+In this tutorial, we introduce the **mlr3fselect** package by comparing feature selection methods on the titanic disaster data set. 
+The objective of feature selection is to enhance the interpretability of models, speed up the learning process and increase the predictive performance.
+
+## Titanic data set
+
+The [Titanic data set](https://www.kaggle.com/c/titanic/data) contains data for 887 Titanic passengers, including whether they survived when the titanic sank.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
The [Titanic data set](https://www.kaggle.com/c/titanic/data) contains data for 887 Titanic passengers, including whether they survived when the Titanic sank. 
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+## Introduction
+
+In this tutorial, we introduce the **mlr3fselect** package by comparing feature selection methods on the titanic disaster data set. 
+The objective of feature selection is to enhance the interpretability of models, speed up the learning process and increase the predictive performance.
+
+## Titanic data set
+
+The [Titanic data set](https://www.kaggle.com/c/titanic/data) contains data for 887 Titanic passengers, including whether they survived when the titanic sank. 
+Our goal will be to predict the survival of the titanic passengers.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
Our goal will be to predict the survival of the Titanic passengers. 
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+## Titanic data set
+
+The [Titanic data set](https://www.kaggle.com/c/titanic/data) contains data for 887 Titanic passengers, including whether they survived when the titanic sank. 
+Our goal will be to predict the survival of the titanic passengers. 
+
+After loading the data set from the `mlr3data` package, we impute the missing age values with the median age of the passengers, set missing embarked values to `s` and remove `character` features.
+We could use feature engineering to create new features from the `character` features, however we want to focus on feature selection in this tutorial.
+
+In addition to the `survived` column, the reduced data set contains for each passenger the following attributes:
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
In addition to the `survived` column, the reduced data set contains the following attributes for each passenger:
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+* `age` Age 
+* `sex` Sex 
+* `sib_sp` Number of siblings/ spouses aboard
+* `parch` Number of parents/ children aboard
+* `fare` Fair paid
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
* `fare` Ammount paid for the ticket
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+resampling = rsmp("cv", folds = 3)
+measure = msr("classif.ce")
+
+resampling$instantiate(task)
+```
+
+## Classes
+
+The `FSelectInstanceSingleCrit` class specifies a general feature selection scenario. 
+It includes the  `ObjectiveFSelect` object that encodes the black box objective function which is optimized by a feature selection algorithm.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
It includes the `ObjectiveFSelect` object that encodes the black box objective function which is optimized by a feature selection algorithm. 
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+## Classes
+
+The `FSelectInstanceSingleCrit` class specifies a general feature selection scenario. 
+It includes the  `ObjectiveFSelect` object that encodes the black box objective function which is optimized by a feature selection algorithm. 
+The evaluated feature sets are stored in an `ArchiveFSelect` object. 
+The archive provides a method for querying the best performing feature set.
+
+The `Terminator` subclasses determine when to stop the feature selection. 
+In this example we choose a terminator that stops the feature selection after 10 seconds. 
+The sugar functions `trm()` and `trms()` can be used to retrieve terminators from the `mlr_terminators` dictionary.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:96]:
```suggestion
The shortcut functions `trm()` and `trms()` can be used to retrieve terminators from the `mlr_terminators` dictionary.
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+terminator = trm("run_time", secs = 10)
+FSelectInstanceSingleCrit$new(
+  task = task, 
+  learner = learner, 
+  resampling = resampling, 
+  measure = measure, 
+  terminator = terminator)
+```
+
+The `FSelector` subclasses describe the feature selection strategy. The sugar function `fs()` can be used to retrieve feature selection algorithms from the `mlr_fselectors` dictionary.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:110]:
```suggestion
The `FSelector` subclasses describe the feature selection strategy. The shortcut function `fs()` can be used to retrieve feature selection algorithms from the `mlr_fselectors` dictionary.
```

Comment (jakob-r) [reply _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:110]: Maybe mention that `fs()` lists all available methods?

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+```{r}
+mlr_fselectors
+```
+
+## Random search
+
+Random search randomly draws feature sets and evaluates them in batches.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:118]: maybe mention that each feature has a chance of 0.5 to be drawn and that small feature sets and big feature sets are unlikely?

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+The `FSelector` subclasses describe the feature selection strategy. The sugar function `fs()` can be used to retrieve feature selection algorithms from the `mlr_fselectors` dictionary.
+
+```{r}
+mlr_fselectors
+```
+
+## Random search
+
+Random search randomly draws feature sets and evaluates them in batches. 
+We retrieve the `FSelectRandomSearch` class with the `fs()` sugar function and choose the `evals` terminator.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:119]:
```suggestion
We retrieve the `FSelectRandomSearch` class with the `fs()` function and choose the `evals` terminator. 
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+  resampling = resampling, 
+  measure = measure, 
+  terminator = terminator)
+fselector = fs("random_search", batch_size = 5)
+```
+
+The feature selection is started by passing the `FSelectInstanceSingleCrit` object to the `$optimize()` method of `FSelectorRandomSearch` which generates the feature sets. 
+These features set are internally passed to the `$eval_batch()` method of `FSelectInstanceSingleCrit` which evaluates them with the objective fuction and stores the results in the archive. 
+This general interaction between the objects of **mlr3fselect** stays the same for the different feature selection methods.
+However, the way how new feature sets are generated differs depending on the chosen `FSelector` subclass.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:136]:
```suggestion
However, the way how new feature sets are generated is determined by the `FSelector` subclass.
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+
+We retrieve the best performing feature set with 
+
+```{r}
+instance$result
+```
+
+## Sequential forward selection 
+
+We try sequential forward selection. We chose the `stagnation` terminator that stops the feature selection if the predictive performance does not increase anymore.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:162]:
```suggestion
We try sequential forward selection. We chose the `stagnation` terminator that stops the feature selection once the classification error does not decrease anymore.
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+
+```{r}
+fselector$optimization_path(instance)
+```
+
+## Recursive feature elimination
+
+Recrusive feature elemiation utilizes the `$importance()` method of learners.
+In each iteration the feature(s) with the lowest importance score is droped. 
+We choose the non-recruvie algorithm (`recursive = FALSE`) which calculates the feature importance once on the complete feature set.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd:187]:
```suggestion
We choose the non-recursive algorithm (`recursive = FALSE`) which calculates the feature importance once on the complete feature set. 
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+
+We access the results.
+
+```{r}
+as.data.table(instance$archive)[, 1:8]
+```
+
+## Nested resampling
+
+It is a common mistake to report the predictive performance estimated on resampling sets during the feature selection as the performance that can be expected from the model.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
It is a common mistake to report the predictive performance estimated on resampling sets during the feature selection as the performance that can be expected from the combined feature selection and model training.
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+The repeated evaluation of the model might leak information about the test sets into the model and thus leads to overfitting and over-optimistic performance results. 
+[Nested resampling](https://mlr3book.mlr-org.com/nested-resampling.html) uses an outer and inner resampling to separate the feature selection from the performance estimation of the model.
+We can use the `AutoFselector` class for running nested resampling.
+The `AutoFselector` wraps a `Learner` and argument it with an automatic feature selection.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
The `AutoFselector` essantially combines a given `Learner` and feature selection method into a `Learner` with internal automatic feature selection.
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+This is our inner resampling loop.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
The inner resampling loop that is used to determine the best feature set is conducted internally each time the `AutoFSelector` `Learner` object is trained.
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+```
+
+Next, we want to compare the predictive performances estimated on the outer resampling to the inner resampling.
+Significantly lower predictive performances on the outer resampling indicate that the models with the optimized feature sets overfit the data.
+
+```{r}
+rr$score()
+```
+
+We check if the feature sets are stable that were selected in the inner resampling.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
We check whether the feature sets that were selected in the inner resampling are stable.
```

File: _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+We check if the feature sets are stable that were selected in the inner resampling. 
+The selected featute sets should not differ too much.
+We might observe unstable models in this example because the small data set and the low number of resampling iterations might introduces too much randomness. 
+Usually, we aim for the selection of similiar feature sets for all outer training sets.
+
+```{r}
+do.call(rbind, lapply(rr$learners, function(x) x$fselect_result))
+```
+
+The archive of the `AutoFSelector` gives us all evaluated hyperparameter configurations with the associated predictive performances.
```
Comment (jakob-r) [inline _posts/2020-09-14-mlr3fselect-basic/mlr3fselect-basic.Rmd]:
```suggestion
The archive of the `AutoFSelector` gives us all evaluated hyperparameter configurations (i.e.
 feature sets) with the associated predictive performances.
```

**Review by jakob-r — COMMENTED** [review]
Quite good already, small corrections!
